app/controllers/concerns/grouping_helper.rb

```diff
@@ -0,0 +1,68 @@
+module GroupingHelper
+
+  class GroupedIncumbency
+    attr_accessor :start_date, :end_date, :nodes, :grouped_by, :type
+  end
+
+  # Returns array of grouped, ungrouped and unknown objects
+  def self.group(data, category, property)
+    grouped_data = group_data(data, category, property)
+    create_sorted_array(grouped_data)
+  end
+
+  # Returns hash with grouped keys
+  # e.g. { 12345678 => [...], 23456789 => [...], "UNKNOWN" => [...] }
+  def self.group_data(data, category, property)
+    data.group_by { |data_point| data_point.try(category).try(property) || 'UNKNOWN' }
+  end
+
+  # Create array with all UNKNOWN nodes and create array with all ungrouped data
+  def self.create_sorted_array(data_hash)
+    sorted_array = []
+    data_hash.each do |key, value|
+      if key == 'UNKNOWN' || value.length == 1
+        sorted_array << value
+      else
+        sorted_array << separate_grouped_data(data_hash, key)
+      end
+    end
+    sorted_array.flatten
+  end
+
+  def self.separate_grouped_data(data_hash, key)
+    grouped = []
+
+    grouped_object = GroupingHelper::GroupedIncumbency.new
+    grouped_object.grouped_by = key
+    grouped_object.nodes = data_hash[key]
+
+    start_date = nil
+    end_date = nil
+    current_node = nil
+
+    grouped_object.nodes.each do |node|
+      # Find node that represents current role
+      # TODO: Might need to cater for multiple current nodes - eg. multiple current memberships and roles within a committee
+      current_node = node if node.try(:end_date).nil? || node.end_date.nil?
+      # Find earliest start date and latest end date
+      start_date = node.start_date if start_date.nil? || node.start_date < start_date
+      end_date = node.end_date if !current_node && (end_date.nil? || node.end_date > end_date)
+    end
+
+    # Set properties of the object
+    # TODO: Might need to adjust to allow for multiple roles roles and memberships of a committee
+    grouped_object.type = grouped_object.nodes.first.type
+    grouped_object.start_date = start_date
+    grouped_object.end_date = current_node.nil? ? end_date : nil
+
+    # Add sorted object to the array
+    sort_grouped_nodes_by_date(grouped_object, current_node)
+    grouped << grouped_object
+  end
+
+  def self.sort_grouped_nodes_by_date(grouped_object, current_node)
```

Comment:
do you need to pass current_node in this instance? Could you not simplify by taking it straight from grouped_object?